### PR TITLE
Payment Requests

### LINKED ISSUE #3453: ThaliaPleasePay
### What?
Payment Request, or ThaliaPleasePay (`TTP'tje`)

### Why?
Invoices are annoying, they require administrative work on both the sending and receiving end. This could be made much easier by using Thalia(Please)Pay an admin would get a nice overview of which request has been payed through a Payment for a Payable (could be any type of payment). And a payer would be able to use Thalia Pay instead of manually transferring the payment.

### How?
A new concrexit app, `paymentrequests` is needed for this, it will depend on `payments`

In admin there will be an entry per `TTP'tje`, which contains a few display fields matching the payment, namely `amount`, `topic`, `notes`. And an inline row per `Member` that has been requested to pay, this is a `Payment Request` and is a `Payable` object. It also has a status, it can be `Pending`, `Paid`, `Cancelled` (essentially a removed payment request, and therefor a separate flag on the class).

The create form will have text fields for `amount`, `topic` and `notes`, and an inline where you can search for and select members to whom the payment request apply (this may also be a single Member). Upon saving this request will just be saved as a concept (_iff_ at least one Member is selected). It will now also be displayed in the Admin page with a `[Concept]` label. Upon sending this will change to `[Pending]`, and once there are no more `Pending` payments linked to the `TTP` the status will change to `[Done]`.

Upon sending all Members linked to the `TTP` through a Payable will be notified of the request, this could be through a Push Notification and/or an e-mail.

### PR DESCRIPTION
**Pull request by mark-boute**
Closes #3453 and #941.

### Implementation

- We will extend the payment models with a PaymentRequest class, this model will store information about the request and the payment itself, it is also a Payable object. Each payed request will become immutable.

- We will add an AdminView which shows all requests, but the default filter only shows 'pending' (unpaid) requests. 

- We will add an AdminView for editing pending requests.

- There will be an AdminForm for adding new requests, here you can fill in details, and select any number of PaymentUsers (must be at least 1!) to whom the request apply. 

- !TODO think of how we would like to handle late-cancellation fines to happen with the payment-request system

- After creating a new request the user should be notified, we will do this by e-mail and by pushnotification. An admin can send-on-save or do it manually later

### Tasklist

- [x]  setup model
- [ ]  write tests for model
- [ ]  admin views
- [ ]  admin view tests
- [ ]  admin form
- [ ]  admin form tests
- [ ]  notification time 
- [ ]  email tests
- [ ]  a connection between fines and service
- [ ]  connection tests
- [ ]   (optional) reminder date

### Notes
- How can we ensure that Users with an expired Membership can still pay (both before the PaymentRequest is created and after if they didn't pay yet but their membership did expire)?

### CODE REVIEW

**Comment by DeD1rk** [comment]
> Upon creating a new request the user should be notified, we will do this by e-mail and by pushnotification

This should probably be configurable or a separate manual action. I think it's useful if you can also silenty create a PaymentRequest, send it later, or resend it. 
Also, we might want to wait with adding pushnotifications until the app supports paying with a deeplink (very similar to how the app supports paying sales orders).

File: website/payments/models.py [hunk 1]
```diff
@@ -480,3 +481,78 @@ def __str__(self):
 
     class Meta:
         ordering = ("created_at",)
+
+
+class PaymentRequest(Payable):
+    requester = models.OneToOneField(Member, on_delete=models.CASCADE)
```
Comment (DeD1rk) [inline website/payments/models.py:487]:
```suggestion
    created_by = models.OneToOneField(Member, on_delete=models.PROTECT)
```

File: website/payments/models.py [hunk 2]
```diff
@@ -480,3 +481,78 @@ def __str__(self):
...
+
+    request_timestamp = models.DateTimeField(
```
Comment (DeD1rk) [inline website/payments/models.py:489]:
```suggestion
    created_at = models.DateTimeField(
```

File: website/payments/models.py [hunk 3]
```diff
@@ -480,3 +481,78 @@ def __str__(self):
...
+        verbose_name=_("request timestamp"),
+        blank=False,
```
Comment (DeD1rk) [inline website/payments/models.py:491]:
```suggestion
        auto_now_add=True,
        editable=False,
        blank=False,
```

File: website/payments/models.py [hunk 4]
```diff
@@ -480,3 +481,78 @@ def __str__(self):
...
+        null=False,
+    )
+
+    payer = models.OneToOneField(PaymentUser, on_delete=models.CASCADE)
```
Comment (DeD1rk) [inline website/payments/models.py:495]:
```suggestion
    payer = models.ForeignKey(PaymentUser, on_delete=models.PROTECT)
```

File: website/payments/models.py [hunk 5]
```diff
@@ -480,3 +481,78 @@ def __str__(self):
...
+
+    payment_timestamp = models.DateTimeField(
```
Comment (DeD1rk) [inline website/payments/models.py:497]: I think you can just use `payment.created_at`

Comment (DeD1rk) [reply website/payments/models.py:497]: But a `payment = OneToOneField(Payment...` is needed.

File: website/payments/models.py [hunk 6]
```diff
@@ -480,3 +481,78 @@ def __str__(self):
...
+        null=False,
+    )
+
+    notes = models.TextField(
+        max_length=256,
+        verbose_name=_("payment notes"),
+        blank=True,
+        null=True,
+    )
+
```
Comment (DeD1rk) [inline website/payments/models.py:522]: There should probably be some field to decide (for bookkeeping purposes) whether the payable 'has to be paid sometime'? I don't know enough about that though.

File: website/payments/models.py [hunk 7]
```diff
@@ -480,3 +481,78 @@ def __str__(self):
...
+    @property
+    def payed(self):
```
Comment (DeD1rk) [inline website/payments/models.py:524]: Payable interface implementation has to be in a separate class (in a payables.py) to match how all other payables work.

File: website/payments/models.py (hunk 6 above)
Comment (T8902) [reply website/payments/models.py:522]: Should we confirm with roel/nick/someone else that knows about bookkeeping what requirements this field should uphold?

**Comment by mark-boute** [comment]
> > Upon creating a new request the user should be notified, we will do this by e-mail and by pushnotification
> 
> This should probably be configurable or a separate manual action. I think it's useful if you can also silenty create a PaymentRequest, send it later, or resend it. Also, we might want to wait with adding pushnotifications until the app supports paying with a deeplink (very similar to how the app supports paying sales orders).

Changed to: `After creating a new request the user should be notified, we will do this by e-mail and by pushnotification. An admin can send-on-save or do it manually later`

**Pull request closed without merging**